src/libstd/panic.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Panic support in the standard library
+
+#![unstable(feature = "std_panic", reason = "module recently added",
+            issue = "27719")]
+
+use cell::UnsafeCell;
+use ops::{Deref, DerefMut};
+use ptr::Unique;
+use sync::{Mutex, RwLock};
+use sys_common::unwind;
+use thread::Result;
+
+/// A marker trait which represents "panic safe" types in Rust.
+///
+/// This trait is implemented by default for many types and behaves similarly in
+/// terms of inference of implementation to the `Send` and `Sync` traits. The
+/// purpose of this trait is to encode what types are safe to cross a `recover`
+/// boundary with no fear of exception safety. For more information about
+/// exception safety and how it applies to Rust, see an [associated RFC][rfc].
+///
+/// [rfc]: https://github.com/rust-lang/rfcs/blob/master/text/1236-stabilize-catch-panic.md
+///
+/// Types such as `&mut T` and `&RefCell<T>` are examples which are **not**
+/// panic safe. The general idea is that any mutable state which can be shared
+/// across `recover` is not panic safe by default, but types like `&Mutex<T>`
+/// are indeed panic safe because they implement poisoning by default.
+///
+/// Note that it is not intended that most types need to worry about this trait.
+/// It is only used as a bound on the `recover` function and it is largely
+/// advisory as it is not an unsafe trait. The `AssertPanicSafe` wrapper struct
+/// in this module can be used to force this trait to be implemented for any
+/// closed over variables passed to the `recover` function (more on this below).
+#[unstable(feature = "recover", reason = "recent addition", issue = "27719")]
+#[rustc_on_unimplemented = "the type {Self} may not be safely transferred \
+                            across a recover boundary"]
+pub trait PanicSafe {}
```

Comment:
One rationale is that, unlike the existing OIBITs, this should be a very unusual marker to manually talk about in a type. But it could reasonably live in `marker` as well.